gensim/models/keyedvectors.py

```diff
@@ -746,27 +806,28 @@ def most_similar(
             positive = [positive]
 
         # add weights for each key, if not already present; default to 1.0 for positive and -1.0 for negative keys
-        positive = [
-            (item, 1.0) if isinstance(item, KEY_TYPES + (ndarray,))
-            else item for item in positive
-        ]
-        negative = [
-            (item, -1.0) if isinstance(item, KEY_TYPES + (ndarray,))
-            else item for item in negative
-        ]
+        keys, weight = [], []
+        for item in positive:
+            if isinstance(item, KEY_TYPES + (ndarray,)):
+                keys.append(item)
+                weight.append(1.0)
+            else:
+                keys.append(item[0])
+                weight.append(item[1])
+        for item in negative:
+            if isinstance(item, KEY_TYPES + (ndarray,)):
+                keys.append(item)
+                weight.append(-1.0)
+            else:
+                keys.append(item[0])
+                weight.append(item[1])
```

Comment:
This seems a bit verbose (& may need further refactoring to gel with PR #3000 also pending).  It also looks like it's expending effort to retain an (undocumented) option of specyfing the `postiive` or `negative` list with `(key, weight)` tuples - which may be interesting, and even in use somewhere, but if it's a live/maintained capability should be mentioned in doc-comment. 

Maybe the two loops can be merged by concatenating `positive` & `negative` earlier? Maybe the repetitive branches can be minimized by pre-composing `weights`?

Reply:
We can concatenate the `positive` and `negative` to reduce for loop, but I am not sure how to minimize the if, else branches as we need to check for every element if it's of type `KEY_TYPES` or `tuple`.

Reply:
I am thinking of refactoring the above one like this to reduce one `for` loop. Is there any other way I can further improve this one?
```py
keys = []
weight = np.concatenate((np.ones(len(positive)), -1.0*np.ones(len(negative))))
for idx,item in enumerate(positive+negative):
    if isinstance(item, KEY_TYPES + (ndarray,)):
        keys.append(item)
    else:
        keys.append(item[0])
        weight[idx] = item[1]
```